File: java/Hot100/Heap/MedianFinder.java
package Hot100.Heap;

import java.util.PriorityQueue;

class MedianFinder {
    private PriorityQueue<Integer> maxHeap; // 维护较小的一半元素（大顶堆）
    private PriorityQueue<Integer> minHeap; // 维护较大的一半元素（小顶堆）

    public MedianFinder() {
        // 大顶堆：自定义 Comparator 实现降序
        maxHeap = new PriorityQueue<>((a, b) -> Integer.compare(b, a));
        // 小顶堆：默认升序
        minHeap = new PriorityQueue<>();
    }

    public void addNum(int num) {
        // 1. 先入大顶堆，将大顶堆的最大值筛选并压入小顶堆
        maxHeap.offer(num);
        minHeap.offer(maxHeap.poll());

        // 2. 平衡数量：确保 maxHeap 的元素数量不少于 minHeap
        if (minHeap.size() > maxHeap.size()) {
            maxHeap.offer(minHeap.poll());
        }
    }

    public double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.peek();
        }
        return (maxHeap.peek() + minHeap.peek()) / 2.0;
    }
}

File: python/heap/median_finder.py
import heapq


class MedianFinder:
    def __init__(self):
        # max_heap 维护较小的一半元素（存相反数以实现大顶堆）
        self.max_heap = []
        # min_heap 维护较大的一半元素（小顶堆）
        self.min_heap = []

    def add_num(self, num: int) -> None:
        # 1. 先压入 max_heap，再将 max_heap 中的最大值弹出并放入 min_heap
        #    利用 -num 实现大顶堆逻辑
        val = -heapq.heappushpop(self.max_heap, -num)
        heapq.heappush(self.min_heap, val)

        # 2. 保证 max_heap 的数量不少于 min_heap（允许 max_heap 比 min_heap 多 1 个）
        if len(self.min_heap) > len(self.max_heap):
            heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))

    def find_median(self) -> float:
        if len(self.max_heap) > len(self.min_heap):
            return float(-self.max_heap[0])
        return (-self.max_heap[0] + self.min_heap[0]) / 2.0